client/question/services/FeedbackGeneratorService.js

```diff
@@ -263,34 +328,75 @@ tie.factory('FeedbackGeneratorService', [
      * @param {string|null} outputFunctionName If needed, is string of the
      *    name for the function needed to process/format the user's output.
      * @param {CorrectnessTest} TestCase
+     * @param {number} testCaseIndex
      * @param {*} observedOutput Actual output for running user's code.
+     * @param {string} testSuiteId
      * @returns {Feedback}
      * @private
      */
     var _getCorrectnessTestFeedback = function(
-      outputFunctionName, testCase, observedOutput) {
+        outputFunctionName, testCase, testCaseIndex, observedOutput,
+        testSuiteId) {
+      var testCaseKey = testSuiteId + '-' + testCaseIndex;
       var allowedOutputExample = testCase.getAnyAllowedOutput();
       var feedback = FeedbackObjectFactory.create(
         FEEDBACK_CATEGORIES.INCORRECT_OUTPUT_FAILURE);
-      // TODO(eyurko): Add varied statements for when code is incorrect.
-      feedback.appendTextParagraph('Your code produced the following result:');
-      if (outputFunctionName) {
-        feedback.appendCodeParagraph(
-          'Input: ' + _jsToHumanReadable(testCase.getInput()) + '\n' +
-          ('Output (after running ' + outputFunctionName + '): ' +
-            _jsToHumanReadable(observedOutput)));
-        feedback.appendTextParagraph(
-          'However, the expected output of ' + outputFunctionName + ' is:');
-      } else {
-        feedback.appendCodeParagraph(
-          'Input: ' + _jsToHumanReadable(testCase.getInput()) + '\n' +
-          'Output: ' + _jsToHumanReadable(observedOutput));
-        feedback.appendTextParagraph('However, the expected output is:');
+      // Check if new / next test suite
+      if (testSuiteId != previousTestSuiteId) {
+        previousTestSuiteId = testSuiteId;
+        // Catch regressions
+        if (correctnessTestStates.hasOwnProperty(testCaseKey)) {
```

Comment:
Just to check, is the logic here correct? For example, if the user has passed suite 1 and 2 without getting any feedback, and subsequently regresses, would there even be an entry for those test cases in correctnessTestStates?